### PR TITLE
Added VisitDirectDeps* methods to SingletonContext.

### PR DESCRIPTION
**Pull request by skvadrik**
A an example use case is generating the graph of module
dependencies, where modules are vertices and dependencies
are edges.

The patch was tested with a local implementation of such
a module in the Soong build system. The graph was generated
for all modules in the AOSP tree.

### CODE REVIEW

File: singleton_ctx.go [hunk 1]
```diff
@@ -93,6 +93,26 @@ type SingletonContext interface {
...
+	// VisitDirectDeps calls visit for each direct dependency of the Module.
+	// If there are multiple direct dependencies on the same module visit will
+	// be called multiple times on that module and OtherModuleDependencyTag
+	// will return a different tag for each.
+	//
+	// The Module passed to the visit function should not be retained outside
+	// of the visit function, it may be invalidated by future mutators.
+	VisitDirectDeps(module Module, visit func(Module))
+
+	// VisitDirectDepsIf calls pred for each direct dependency of th Module,
```
Comment (colincross) [inline singleton_ctx.go]: th -> the

File: singleton_ctx.go [hunk 2]
```diff
@@ -93,6 +93,26 @@ type SingletonContext interface {
...
+	// and if pred returns true calls visit.  If there are multiple direct
+	// dependencies on the same module pred and visit will be called multiple
+	// times on that module and OtherModuleDependencyTag will return a different
+	// tag for each.
+	//
+	// The Module passed to the visit function should not be retained outside
+	// of the visit function, it may be invalidated by future mutators.
+	VisitDirectDepsIf(module Module, pred func(Module) bool,
```
Comment (colincross) [inline singleton_ctx.go]: go uses a soft 100 character line length limit, you can leave this on one line.

File: singleton_ctx.go [hunk 3]
```diff
@@ -289,6 +309,18 @@ func (s *singletonContext) VisitAllModulesIf(pred func(Module) bool,
 	s.context.VisitAllModulesIf(pred, visit)
 }
 
+func (s *singletonContext) VisitDirectDeps(module Module,
```
Comment (colincross) [inline singleton_ctx.go]: same

File: singleton_ctx.go [hunk 4]
```diff
@@ -289,6 +309,18 @@ func (s *singletonContext) VisitAllModulesIf(pred func(Module) bool,
...
+	visit func(Module)) {
+
+	s.context.VisitDirectDeps(module, visit)
+}
+
+func (s *singletonContext) VisitDirectDepsIf(module Module,
```
Comment (colincross) [inline singleton_ctx.go]: same

File: go.mod [hunk 5]
```diff
@@ -1 +1,3 @@
 module github.com/google/blueprint
+
```
Comment (colincross) [inline go.mod]: Can you undo the changes to this file?

Comment (skvadrik) [reply go.mod]: Eh, that was by accident.

**Comment by skvadrik** [comment]
@colincross All comments addressed, thanks for reviewing. I force-pushed the changed commit.
